### ISSUE TITLE
The Syntax of Import and Use

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#2512]
I think the syntax between java imports and flix uses could match closer without causing confusion or ambiguity.
```
namespace Java/Lang/Object {
    pub def toString(_x: a): String = ???
}

def main(_args: Array[String]): Int32 & Impure =
    import java.lang.Object. toString() as _objString1 ;
    use    Java/Lang/Object.{toString   => _objString2};
    println("done"); 0
```
* `as` in the java import could be `=>` to align the two and avoid the naming collision with casting
* The namespace delimiter could be `/` in both cases. It is always clear when a java class is referenced, either by `import` or `##<something>`

These two changes would result in
```
import java/lang/Object. toString() => _objString1 ;
use    Java/Lang/Object.{toString   => _objString2};
```

**Comment by mlutze** [comment]
I'm starting to work on this, but I'd like some consensus before making the changes.

### Proposal 1 (from @JonathanStarup's comment):
We get rid of the `import` keyword and instead use `##` prefix to distinguish Java imports and Flix imports:
```scala
    use ##java.lang.Object.toString(); // instance method
    use ##java.lang.Object.toString() as objectString; // instance method with renaming
    use new ##java.lang.Object() as newObject; // instance constructor
    use ##java.util.Arrays:sort(Array[Char]); // class method
    use ##java.util.Arrays:sort(Array[Char]) as sort; // class method with renaming
    use get ##java.util.ArrayList.length as length; // get instance field (fake example)
    use set ##java.util.ArrayList.length as setLength; // set instance field (fake example)
    use get ##java.lang.System:out as systemOut; // get class field
    use set ##java.lang.System:out as setSystemOut; // set class field
```
### Proposal 2:
Replace `as` with `=>`
```scala
    import get java.util.ArrayList.length => length; // get instance field (fake example)
    import set java.util.ArrayList.length => setLength; // set instance field (fake example)
    import get java.lang.System:out => systemOut; // get class field
    import set java.lang.System:out => setSystemOut; // set class field
```

### Proposal 3:
Support multiple imports per namespace
```scala
    import java.lang.Object.{toString(), hashCode()}; // instance method
    import java.lang.Object.{toString() as objectString, hashCode() as hash}; // instance method with renaming
    import java.util.Arrays:{sort(Array[Char]), sort(Array[Int])}; // class method
    import java.util.Arrays:{sort(Array[Char]) as sort, sort(Array[Int]) as sort2}; // class method with renaming
    import get java.util.ArrayList.{length as length, capacity as capacity}; // get instance field (fake example)
    import set java.util.ArrayList.{length as setLength, capacity as setCapacity}; // set instance field (fake example)
    import get java.lang.System:{out as systemOut, in as systemIn}; // get class field
    import set java.lang.System:{out as setSystemOut, in as setSystemIn}; // set class field
```

### Proposal 4:
Don't require renaming for `get` and `set`
```scala
    import get java.util.ArrayList.length; // get instance field (fake example)
    import set java.util.ArrayList.length; // set instance field (fake example)
    import get java.lang.System:out; // get class field
    import set java.lang.System:out; // set class field
```

### Proposal 5:
Use `/` for Java namespace delimiters
```scala
    import java/lang/Object.toString(); // instance method
    import java/lang/Object.toString() as objectString; // instance method with renaming
    import new java/lang/Object() as newObject; // instance constructor
    import java/util/Arrays:sort(Array[Char]); // class method
    import java/util/Arrays:sort(Array[Char]) as sort; // class method with renaming
    import get java/util/ArrayList.length as length; // get instance field (fake example)
    import set java/util/ArrayList.length as setLength; // set instance field (fake example)
    import get java/lang/System:out as systemOut; // get class field
    import set java/lang/System:out as setSystemOut; // set class field
```

### Other thoughts while we're here

What do we think of the `:` syntax for distinguishing static vs. instance methods/fields? We could get rid of this be treating all methods as functions, so we end up with `import java.util.List.length(##java.util.List)` but that might end up being annoyingly verbose.

Anything I missed?

These are all orthogonal to each other.

**Comment by JonathanStarup** [comment]
Im a fan of 1-4 :) I think it improves the cohesion of "I want to use other code". I remember use and import as being treated  quite differently the the parser which might be confusing when they use the same keyword.

I think for proposal 5 with the `/` delimiter we could also make everything upper case? or is there a possibility that `java.lang.Object` and `Java.Lang.Object` are different things? maybe inner classes use it.

Does anything else use : for static? It doesn't associate for me but it looks nice enough.

**Comment by mlutze** [comment]
I think lowercase namespaces in Java are only convention, so lower and upper namespaces are distinct.

**Comment by JonathanStarup** [comment]
Yeah okay I see, so that can't be changed. I think maybe 5 is the most controversial and if they have a casing difference it is probably nicer just to keep the `.` separator.
I was thinking about the [map entry in java](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html). with a `/` seperator would that be `java/util/Map.Entry.getKey()`?

**Comment by mlutze** [comment]
Yes, I guess with inner classes it is tricky. Currently I think we do something with `$`. I'll keep the delimiter as it is for now and go ahead with 1-4.

**Comment by magnus-madsen** [comment]
What is the problem we are trying to solve? :) I can see a few:

- Dealing with purity directly in an import.
- Making imports more easy to use.
- Making use and imports the same.

I would say that the (3) is a non-goal because these mechanisms exists for different reasons. Moreover, I would argue that the import mechanism should closely mirror Java to make it easier to use. We should always keep in mind that there are Flix names and Java names and these are not the same. Trying to make them the same is not good because Flix aims to be richer than Java and different than Java. We cannot expect a clean mapping between the two. 

I should remark that Scala switched from `import xyz.{foo => bar}` to `import xyz.{foo as bar}` like in Python.

I should also remark that we lack a qualified import like Haskell.

Looking over the proposals I don't really see any real gain. I would say we should make the import mechanism more like the Java one. For example, we could replace the weird use of `:` with `static` e.g.:

```
import static Foo.Bar.baz
import static get Foo.Bar.baz
```

It would be nice if we could deal with fields in a better way. I don't know how.

It would also be nice if we could have purity directly on the import. Perhaps:

```
import java.lang.String.toLowerCase() as & Pure
```

But then we would have 

```
import java.lang.String.toLowerCase() as lower as & Pure
```

so I can see the argument for proposal 2, renaming `as` to `=>`.

So in summary, I suggest three changes:

- use `static` to differential static vs. non-static.
- replace `as` with `=>`
- allow an `& Pure` annotation after the whole thing.

**Comment by JonathanStarup** [comment]
> I would say that the (3) is a non-goal because these mechanisms exists for different reasons.

I agree that I don't think its a goal that they should be the same but I think of it as a part of making it easier to use/remember (2).

**Comment by magnus-madsen** [comment]
> I agree that I don't think its a goal that they should be the same but I think of it as a part of making it easier to use/remember (2).

I think that is why I would go for the updated import syntax. It is what a programmer would remember from Java. (In fact modulo the last part they could even test it in Java). Also, of course by using `=>` everything would be more regular.

**Comment by magnus-madsen** [comment]
If we implement support for these three items:

- [ ] use static to differential static vs. non-static.
- [ ] replace as with =>.
- [ ] allow return type and effect annotations "after the whole thing".

(without breaking the current syntax) then I can ask one of the student programmers to refactor the standard library and then we can remove support at the end.

**Comment by JonathanStarup** [comment]
I still think it would be nice to call them both `use` since they are both references to external code and the `##` type tag differentiates the two clearly.
> We should always keep in mind that there are Flix names and Java names and these are not the same.

I think this is why `import java.lang.String...` feels weird because its the only place `##` isn't used to indicate a java name

Right now imports can't be used at top-level which feels a bit arbitrary?

**Comment by magnus-madsen** [comment]
> I think this is why `import java.lang.String...` feels weird because its the only place `##` isn't used to indicate a java name

Yes. It would be nice if we could just write Java types in a more natural way, but not sure the grammar allows that.

> Right now imports can't be used at top-level which feels a bit arbitrary?

This is to discourage broad use of import.

**Comment by magnus-madsen** [comment]
@mlutze Want to chime in?

I was going to add support for the `static` keyword and for `=>` (or ask one of you) and then have one of the student programmer start on the refactoring.

**Comment by mlutze** [comment]
I tried unifying imports and uses while working on this earlier, but it turned into a major pain since the mechanism we use to deal with them is so different: for `use` we just add to the namespace, but for `import` we build a lambda that calls the function. So I'm less inclined now to want to unify their syntax.

It might still be nice to be able to declare multiple imports in the same line, but I agree it's not as important as the rest.

I think allowing both type and effect casts at the import site could be nice in case a function returns a `LinkedList` but we need to use it with a function that takes a `List` for example.

**Comment by magnus-madsen** [comment]
> I tried unifying imports and uses while working on this earlier, but it turned into a major pain since the mechanism we use to deal with them is so different: for `use` we just add to the namespace, but for `import` we build a lambda that calls the function. So I'm less inclined now to want to unify their syntax.

Philosophically speaking it is good to unify things that are instances of the same concept. But JVM names and Flix names are not instances of the same concept. Java has things we don't have: classes, interfaces, abstract classes, inner classes, static inner classes. Flix has things that Java does not: namespaces, functions not associated with classes, type classes, ADTs, etc. We should not unify these mechanisms because there is nothing in common and it will make it harder to grow Flix. Its better to improve the Java mechanism and to try to make it work well with Java and even to emulate Java.

**Comment by mlutze** [comment]
Should we require braces around renamings (as we do with `use`)?

```scala
import java.util.String.{indexOf(String) => myIndexOf} as & Pure
// vs
import java.util.String.indexOf(String) => myIndexOf as & Pure
```

**Comment by magnus-madsen** [comment]
> Should we require braces around renamings (as we do with `use`)?
> 
> ```scala
> import java.util.String.{indexOf(String) => myIndexOf} as & Pure
> // vs
> import java.util.String.indexOf(String) => myIndexOf as & Pure
> ```

Yes, I think so.

Does Java have renaming? Guess not?

**Comment by magnus-madsen** [comment]
@ninaandrup I would like you to:

- [ ] Open `Parser.scala` and find `JvmStaticName`, `StaticMethodOLD`, `GetStaticFieldOLD`, and `PutStaticFieldOLD` and remove them. When you do, this will give you errors in the standard library. You resolve these errors as follows:

If you have:

```
import get java.util.concurrent.TimeUnit:NANOSECONDS as nano;
```

then change it to:

```
import static get java.util.concurrent.TimeUnit.NANOSECONDS as nano;
```

NOTE HOW THE : (COLON) becomes a dot. 

Makes sense?

**Comment by magnus-madsen** [comment]
See also https://github.com/flix/flix/pull/2757/files

**Comment by mlutze** [comment]
Should we do the`=>` syntax change at the same time?

**Comment by magnus-madsen** [comment]
> Should we do the`=>` syntax change at the same time?

I think we do everything stepwise.

**Comment by ninaandrup** [comment]
> @ninaandrup I would like you to:
> 
> * [ ]  Open `Parser.scala` and find `JvmStaticName`, `StaticMethodOLD`, `GetStaticFieldOLD`, and `PutStaticFieldOLD` and remove them. When you do, this will give you errors in the standard library. You resolve these errors as follows:
> 
> If you have:
> 
> ```
> import get java.util.concurrent.TimeUnit:NANOSECONDS as nano;
> ```
> 
> then change it to:
> 
> ```
> import static get java.util.concurrent.TimeUnit.NANOSECONDS as nano;
> ```
> 
> NOTE HOW THE : (COLON) becomes a dot.
> 
> Makes sense?

Makes sense, I will try to do this now.

**Comment by magnus-madsen** [comment]
Thanks to Matt, we can now write the following:

```
    pub def charAt(i: Int32, s: String): Char =
        import java.lang.String.charAt(Int32);
        charAt(s, i) as & Pure
```

as

```
    pub def charAt(i: Int32, s: String): Char =
        import java.lang.String.charAt(Int32): Char & Pure;
        charAt(s, i)
```

While this may not seem like much of a change, it is actually very useful internally, because it means that immediately know that the imported `charAt` is pure. (While the syntax does allow us to omit the return type, I suggest we should not do it by default.)

@ninaandrup I would like to apply the above refactoring through the standard library. There is a lot to do and its not high-priority, so it is something you can work on where there is time or you are stuck (or waiting for feedback etc). I suggest you do one namespace at a time. You could start with `String`. Does that make sense?
